base-server/index.js

```diff
@@ -355,6 +368,46 @@ export class BaseServer {
     this.listenNotes = {}
     bindBackendProxy(this)
 
+    let end = (actionId, queue, queueKey, ...args) => {
+      this.actionToQueue.delete(actionId)
+      if (queue.length() === 0) {
+        this.queues.delete(queueKey)
+      }
+      queue.next(...args)
+    }
+    let undoRemainingTasks = queue => {
+      let remainingTasks = queue.getQueue()
+      if (remainingTasks) {
+        for (let task of remainingTasks) {
+          this.undo(task.action, task.meta, 'error')
+          this.actionToQueue.delete(task.meta.id)
+        }
+      }
+      queue.killAndDrain()
+    }
+    this.on('error', (e, action, meta) => {
+      let queueKey = this.actionToQueue.get(meta?.id)
+      if (!queueKey) {
+        return
+      }
+      let queue = this.queues.get(queueKey)
+      undoRemainingTasks(queue)
+      end(meta.id, queue, queueKey, e)
+    })
+    this.on('processed', (action, meta) => {
+      let queueKey =
+        this.actionToQueue.get(meta?.id) || this.actionToQueue.get(action?.id)
```

Comment:
Why do we need `this.actionToQueue.get(action?.id)`?

Reply:
Do we need it for `logux/undo` actions? Let’s put `this.actionToQueue.get(action?.id)` in `if (action.type === 'logux/undo')` branch to make it clear and avoid bugs if some other action will have `action.id`.

Reply:
Actions are saved to the `actionToQueue` by `meta.id`. `on('processed')` will receive `logux/processed` which [per docs](https://logux.org/guide/concepts/action/#system-actions) has `action.id` equal to the `meta.id` of the processed action.

Actually we don't need `this.actionToQueue.get(meta?.id)` here.